### PR TITLE
Allow cyrilic domain names in email addresses

### LINKED ISSUE #11588: Email validation are not checked cyrillic characters before the @ sign. Also there is no support for Cyrillic domains.
**Describe the bug**
At the php level, the email validation form skips the cyrillic characters before the @ sign. Also, the form has no support for cyrillistic domains.

**How it should work**
Cyrillistic domains are supported by some email providers, such as Yandex or Google (Gmail).
Cyrillic characters should not be before the @ sign, only after it in domain, if the domain is converted to punycode.
For example:
test@site.com - this is a valid email address.
test@сайт.рф - this is a valid email address ( after in conversion to pinycode: test@xn--80aswg.xn--p1ai )
иван@сайт.рф - this is not a valid email address.
иван@site.com - this is not a valid email address too.
It should work that way. 

**To Reproduce**

1. Go to the [Create Account page](http://fo.demo.prestashop.com/en/login?create_account=1)
2. First, we need to disable html5 validation (because we will test only server validation).
For that we need to add 'novalidate' atribute in `<form>` tag. 
3. On the page in the Email field we write: test@сайт.рф
4. Click "Save" button. And we will see the message: 
> Invalid format. 
This is the first problem. The second is cyrillic characters before the @ sign.
5. Now repeat step 2 and enter another email address: тест@site.com
6. After that we will see the message: 
> Fatal error: Uncaught exception 'Swift_RfcComplianceException' with message 'Address in mailbox given [тест@test.site] does not comply with RFC 2822, 3.6.2.'

**Screenshots**
Step2:
Step2
Step3:
Step3
Step6:
Step6

**Additionnal information**
PrestaShop version: from 1.6 to 1.7
PHP version: all

### PR DESCRIPTION
**Pull request by atomiix**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.7.x
| Description?  | Before this change, it was impossible to use email addresses that contains cyrillic characters in the host part of the address.
| Type?         | bug fix
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #11588
| How to test?  | Try to create an account (BO and/or FO) with an email address containing cyrillic characters in the host part (**after** the `@`): it should work. Try with cyrillic characters **before** the `@`, it should **not** work.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/17561)

### CODE REVIEW

**Review by matks — DISMISSED** [review]

File: src/Core/Util/InternationalizedDomainNameConverter.php
Comment (PierreRambaud) [inline src/Core/Util/InternationalizedDomainNameConverter.php:47]: Watch out with `INTL_IDNA_VARIANT_UTS46` it's only available as of ICU 4.6
Because we now use SwiftMailer > 6.2 we should have a look on what SwiftMailer can do 
https://github.com/swiftmailer/swiftmailer/pull/1044

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
Wait for SwiftMailer feedback

File: src/Core/Util/InternationalizedDomainNameConverter.php
Comment (atomiix) [reply src/Core/Util/InternationalizedDomainNameConverter.php:47]: @PierreRambaud I think we're pretty safe because we have the [Symfony polyfill](https://github.com/symfony/polyfill-intl-idn). WDYT?

Comment (PierreRambaud) [reply src/Core/Util/InternationalizedDomainNameConverter.php:47]: You're right :+1:

File: classes/Validate.php
Comment (PierreRambaud) [inline classes/Validate.php]: https://github.com/PrestaShop/PrestaShop/pull/12377#issuecomment-584182583 we should consider using this package instead.

One reason is `FILTER_VALIDATE_EMAIL` will not consider `test@mydomain` as a real address (type of address you can have in your server)

Comment (atomiix) [reply classes/Validate.php]: Done @PierreRambaud

File: composer.json
Comment (Progi1984) [inline composer.json:25]: License of the package ?

**Review by Progi1984 — COMMENTED** [review]
A question ?

File: composer.json
Comment (atomiix) [reply composer.json:25]: MIT ;)

**Comment by Progi1984** [comment]
@atomiix All unit tests are :red_circle:

**Comment by atomiix** [comment]
@Progi1984 yeah I know, I put the issue back in In Progress on the kanban, I'm working on it 😉

**Comment by Progi1984** [comment]
Please use the label WIP in this case. Thanks

**Comment by atomiix** [comment]
@Progi1984 It should be good now. Only have a prettyCI error but it's not coming from the code I pushed 🤔

**Review by Progi1984 — DISMISSED** [review]

File: src/PrestaShopBundle/Resources/config/services/core/grid/grid_data_factory.yml
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/core/grid/grid_data_factory.yml]: This looks like a copy of existing service `prestashop.core.grid.data_factory.email_logs` Why did you need to duplicate it?
I think you can keep/use the original and define your decorator along with it in `src/PrestaShopBundle/Resources/config/services/core/grid/grid_data_factory.yml`

File: classes/Customer.php
Comment (jolelievre) [inline classes/Customer.php]: I'm not sure it's a good idea to instanciate the class at each call of this method..

File: src/Core/Grid/Data/Factory/CustomerGridDataFactoryDecorator.php
Comment (jolelievre) [inline src/Core/Grid/Data/Factory/CustomerGridDataFactoryDecorator.php]: Since you use this class in several services, I guess you could also declare it as a service and inject it?

**Review by jolelievre — COMMENTED** [review]
Little proposal, you could declare a service for `InternationalizedDomainNameConverter`
This would avoid instanciating it in several services, you could also inject it in the handler that use `getEmailForDisplay` for display (which also instanciate the class at each call), the preview handler code would look like something like that:

```php
'email' => $this->converter->emailToUtf8($customer->email),
```

**Review by jolelievre — COMMENTED** [review]
Nice update with service, seems much cleaner now. I have two other feedbacks:
1- regarding the `SwiftMailerValidation` class, since it implements an external interface maybe it should be in the Adapter namespace instead of Core (along with the related exception)

2- this fixes the bug in front office when we create an account, but should we backport this new validation in the CQRS commands as well, for customers created in the back office. We have an `Email` value object that is used in several handlers
The value object already has a private method to validate the format using `filter_var($email, FILTER_VALIDATE_EMAIL)` but I don't think this allows cyrilic characters?

File: src/Core/Domain/Customer/ValueObject/Email.php
Comment (eternoendless) [inline src/Core/Domain/Customer/ValueObject/Email.php]: Missing since, missing deprecation notice

File: src/Core/Util/InternationalizedDomainNameConverter.php
Comment (eternoendless) [inline src/Core/Util/InternationalizedDomainNameConverter.php]: Only classes implementing interfaces should be final

File: src/Core/Domain/Customer/ValueObject/Email.php
Comment (jolelievre) [inline src/Core/Domain/Customer/ValueObject/Email.php]: Since this class is now deprecated, it would be better if we stop using it in the core code, I only found a reference in `GetCustomerForAddressCreation` but maybe there are more

File: src/Core/Domain/ValueObject/Email.php
Comment (PierreRambaud) [inline src/Core/Domain/ValueObject/Email.php]: Is this preg_match really useful?
Because "` azd    @  `"  will return true :thinking:

Comment (atomiix) [reply src/Core/Domain/ValueObject/Email.php]: @PierreRambaud Yeah, it will check if it looks like an email but it's not really trustworthy, maybe just an indication. Do you think we should get rid of this check in this ValueObject?

File: src/Core/Domain/Customer/ValueObject/Email.php
Comment (atomiix) [reply src/Core/Domain/Customer/ValueObject/Email.php]: @jolelievre Yeah, I didn't find this class elsewhere, and even in `GetCustomerForAddressCreation`, it's not really used because phpdoc of the function `getCustomerEmail()` is says that it returns a `Email` but the function declares returning a `string`. I'll update the phpddoc!

File: src/Core/Domain/ValueObject/Email.php
Comment (PierreRambaud) [reply src/Core/Domain/ValueObject/Email.php]: I prefer having no check instead of a useless one :/

**Review by matthieu-rolland — DISMISSED** [review]

**Review by jolelievre — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Review by PierreRambaud — DISMISSED** [review]

**Comment by boubkerbribri** [comment]
@atomiix  please resolve conflicts

**Review by Progi1984 — DISMISSED** [review]

**Review by matks — DISMISSED** [review]

**Comment by Robin-Fischer-PS** [comment]
It's OK for creation and edit (on BO and FO).

But the display of the address is not OK on Advanced Parameters > Team > Employee in BO. Please tell me if this is related to this PR or not : 

pr17561 display mail employee BO

Thanks !

**Comment by atomiix** [comment]
@Robin-Fischer-PS Nice catch, forgot about this place! I'll fix this ;)

**Comment by Robin-Fischer-PS** [comment]
Oddly, there is another bug : The dropdown in order list do not work anymore (it work on the branch).

pr17561 orders dropdown

**Review by matks — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Comment by Robin-Fischer-PS** [comment]
Hi @atomiix !

Adding a customer or employee with Cyrilic mail is OK.

But after adding, it is not possible to login : 

pr17561 login

And when editing the employee/customer, this message is displayed (even if I did not change the mail) : 

pr17561 edit

Sorry :stuck_out_tongue:

File: admin-dev/themes/new-theme/js/app/utils/email-idn.js
Comment (PierreRambaud) [inline admin-dev/themes/new-theme/js/app/utils/email-idn.js]:
```suggestion
const {$} = window;
```

File: admin-dev/themes/new-theme/js/app/utils/email-idn.js
Comment (PierreRambaud) [inline admin-dev/themes/new-theme/js/app/utils/email-idn.js]: To be reusable, `'input[type="email"]'` should be in parameter

File: admin-dev/themes/new-theme/js/app/utils/email-idn.js
Comment (PierreRambaud) [inline admin-dev/themes/new-theme/js/app/utils/email-idn.js:36]: Only the domain needs to be in punycode. This means `parts[1]`   :thinking:

File: classes/form/CustomerForm.php
Comment (PierreRambaud) [inline classes/form/CustomerForm.php]:
```suggestion
        if (!empty($params['email'])) {
```
We don't care about 0, false and empty string :)

File: classes/form/CustomerLoginForm.php
Comment (PierreRambaud) [inline classes/form/CustomerLoginForm.php]: same here

File: admin-dev/themes/new-theme/js/app/utils/email-idn.js
Comment (atomiix) [reply admin-dev/themes/new-theme/js/app/utils/email-idn.js]: @PierreRambaud it's a very specific case actually. This would work only for input[type="email"]. Do you see another use case for this?

File: admin-dev/themes/new-theme/js/app/utils/email-idn.js
Comment (atomiix) [reply admin-dev/themes/new-theme/js/app/utils/email-idn.js:36]: @PierreRambaud right, but as I'm checking that parts[0] won't be changed and the puycode library doesn't touch the local part of an email adresse, I think we should be pretty safe here. [https://github.com/bestiejs/punycode.js#punycodetoasciiinput](https://github.com/bestiejs/punycode.js#punycodetoasciiinput)

Comment (PierreRambaud) [reply admin-dev/themes/new-theme/js/app/utils/email-idn.js:36]: Add a comment to explain maybe?

File: admin-dev/themes/new-theme/js/app/utils/email-idn.js
Comment (PierreRambaud) [reply admin-dev/themes/new-theme/js/app/utils/email-idn.js]: If you have ten forms on your page, you maybe don't want to check them all, they are maybe not using the same validator as you do :thinking:

**Review by jolelievre — APPROVED** [review]
You're becoming quite the expert in Form transformation 😅

**Review by Progi1984 — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Comment by Robin-Fischer-PS** [comment]
It's finally QA :heavy_check_mark: @atomiix !

**Comment by Progi1984** [comment]
Thanks @atomiix

**Pull request merged**
